stackslib/src/chainstate/nakamoto/shadow.rs

```diff
@@ -0,0 +1,991 @@
+    NakamotoBlockObtainMethod, NakamotoChainState, NakamotoStagingBlocksConn,
+    NakamotoStagingBlocksConnRef, NakamotoStagingBlocksTx, SetupBlockResult, SortitionDB,
+    SortitionHandleConn, StacksDBIndexed,
+};
+use crate::chainstate::stacks::boot::RewardSet;
+use crate::chainstate::stacks::db::blocks::DummyEventDispatcher;
+use crate::chainstate::stacks::db::{
+    ChainstateTx, ClarityTx, StacksAccount, StacksChainState, StacksHeaderInfo,
+};
+use crate::chainstate::stacks::miner::{
+    BlockBuilder, BlockLimitFunction, TransactionError, TransactionProblematic, TransactionResult,
+    TransactionSkipped,
+};
+use crate::chainstate::stacks::{
+    CoinbasePayload, Error, StacksTransaction, StacksTransactionSigner, TenureChangeCause,
+    TenureChangePayload, TransactionAnchorMode, TransactionAuth, TransactionPayload,
+    TransactionVersion,
+};
+use crate::clarity::vm::types::StacksAddressExtensions;
+use crate::clarity_vm::clarity::ClarityInstance;
+use crate::clarity_vm::database::SortitionDBRef;
+use crate::net::Error as NetError;
+use crate::util_lib::db::{query_row, u64_to_sql, Error as DBError};
+
+impl NakamotoBlockHeader {
+    /// Is this a shadow block?
+    ///
+    /// This is a special kind of block that is directly inserted into the chainstate by means of a
+    /// consensus rule.  It won't be downloaded or broadcasted, but every node will have it.  They
+    /// get created as a result of a consensus-level SIP in order to restore the chain to working
+    /// order.
+    ///
+    /// Shadow blocks have the high bit of their version field set.
+    pub fn is_shadow_block(&self) -> bool {
+        Self::is_shadow_block_version(self.version)
+    }
+
+    /// Is a block version a shadow block version?
+    pub fn is_shadow_block_version(version: u8) -> bool {
+        version & 0x80 != 0
+    }
+
+    /// Get the signing weight of a shadow block
+    pub fn get_shadow_signer_weight(&self, reward_set: &RewardSet) -> Result<u32, Error> {
+        let Some(signers) = &reward_set.signers else {
+            return Err(ChainstateError::InvalidStacksBlock(
+                "No signers in the reward set".to_string(),
+            ));
+        };
+        let shadow_weight = signers
+            .iter()
+            .fold(0u32, |acc, signer| acc.saturating_add(signer.weight));
+
+        Ok(shadow_weight)
+    }
+}
+
+impl NakamotoBlock {
+    /// Is this block a shadow block?
+    /// Check the header
+    pub fn is_shadow_block(&self) -> bool {
+        self.header.is_shadow_block()
+    }
+
+    /// Verify that if this shadow block has a coinbase, that its VRF proof is consistent with the leader
+    /// public key's VRF key. If there is no coinbase tx, then this is a no-op.
+    pub(crate) fn check_shadow_coinbase_tx(&self, mainnet: bool) -> Result<(), ChainstateError> {
+        assert!(self.is_shadow_block());
+
+        // If this block has a coinbase, then verify that its VRF proof was generated by this
```

Comment:
These comments seem contradictory to what the function does unless I am missing something. It says we check the vrf proof was generated by this block's miner, but then we say we can't verify the VRF proof because the creator of the shadow block isn't a miner? I think its really that we just ensure that there IS a vrf_proof but do not actually do anything else with it. Side Nit: there is a typo "parnet" -> "parent"